### PR TITLE
sql: Add migration to change schema of system.eventlog (simpler version)

### PR DESCRIPTION
**Pull request by a-robinson**
This just swaps out experimental_unique_bytes with uuid_v4 rather than
dealing with changing the type of uniqueID from BYTES to INT.

An alternative to #12158 that cuts out most of the ugliness. This is a truly backwards-compatible change. If you guys agree we should just take this easier route, I'll cook up some unit tests for it.

Again, only the last commit is new.

@bdarnell @vivekmenezes @paperstreet

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12228)

### CODE REVIEW

**Comment by paperstreet** [comment]
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 257 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYntXUCea59EvKPKi2x:-KYntXUCea59EvKPKi2y:brhp0r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/migrations/migrations.go#L257)):*
> ```Go
> 	for retry := retry.Start(retry.Options{MaxRetries: 5}); retry.Next(); {
> 		res := r.sqlExecutor.ExecuteStatements(session, alterStmt, nil)
> 		err = checkQueryResults(res.ResultList, 1)
> ```

Maybe directly check the schema using `information_schema`

---

*[pkg/sql/sqlbase/system.go, line 83 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYnu7YzOPUiIvrJU1qd:-KYnu7YzOPUiIvrJU1qe:bpxb43y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/sql/sqlbase/system.go#L83)):*
> ```Go
> 	// its uniqueID field. This schema will soon be removed from the repo in order
> 	// to allow us to deprecate experimental_unique_bytes().
> 	DeprecatedEventLogTableSchema = `
> ```

A large part of the motivation of migrations (for me) is being able to remove legacy code at some point. Combined with the fact that the migrations themselves have to be idempotent, I would prefer that our precedent for migrations is that (when possible) new clusters are created in such a way that the migrations are essentially no-ops.

Concretely:

New SQL tables are created with the InterleavedFormatVersion, but tables created by old code may have been created with ColumnFamily or Base. Whenever a TableDescriptor is acquired, we run MaybeUpgradeFormatVersion on it. This function first migrates from Base to ColumnFamily (if necessary) then ColumnFamily to Interleaved (if necessary). This is all true today.

I would like to write a system migration that scans every TableDescriptor, runs MaybeUpgradeFormatVersion on it, and writes it back. Then, we should actually be able to remove the two mini-migrations currently in MaybeUpgradeFormatVersion (but leave the hook). This code needs to stick around to power the system migration, but it can be removed from the normal production codepath, simplifying it.

To conclude, I similarly feel EventLog should just be created as we want it in new clusters, while the migration has a big comment explaining what it looked like before. We shouldn't need to keep the old schema around for anything.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Comment by a-robinson** [comment]
Review status: 0 of 15 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 257 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYntXUCea59EvKPKi2x:-KYofgXpd3Pj3nzRHM8W:b791m1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/migrations/migrations.go#L257)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary><blockquote>

Maybe directly check the schema using `information_schema`
</blockquote></details>

And do what if it isn't in the state that we want? Try again? If we're at the point where a SQL statement isn't returning an error even though it didn't succeed, things are pretty broken.

---

*[pkg/sql/sqlbase/system.go, line 83 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYnu7YzOPUiIvrJU1qd:-KYokk27HcoDYeDBX66R:b-cjks80) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/sql/sqlbase/system.go#L83)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary><blockquote>

A large part of the motivation of migrations (for me) is being able to remove legacy code at some point. Combined with the fact that the migrations themselves have to be idempotent, I would prefer that our precedent for migrations is that (when possible) new clusters are created in such a way that the migrations are essentially no-ops.

Concretely:

New SQL tables are created with the InterleavedFormatVersion, but tables created by old code may have been created with ColumnFamily or Base. Whenever a TableDescriptor is acquired, we run MaybeUpgradeFormatVersion on it. This function first migrates from Base to ColumnFamily (if necessary) then ColumnFamily to Interleaved (if necessary). This is all true today.

I would like to write a system migration that scans every TableDescriptor, runs MaybeUpgradeFormatVersion on it, and writes it back. Then, we should actually be able to remove the two mini-migrations currently in MaybeUpgradeFormatVersion (but leave the hook). This code needs to stick around to power the system migration, but it can be removed from the normal production codepath, simplifying it.

To conclude, I similarly feel EventLog should just be created as we want it in new clusters, while the migration has a big comment explaining what it looked like before. We shouldn't need to keep the old schema around for anything.

</blockquote></details>

Agreed. I was going to actually remove this in a later PR (particularly since it was convenient for testing to still have the old schema around), but I suppose it makes the most sense to do it now.

However, as @bdarnell alluded to in https://reviewable.io/reviews/cockroachdb/cockroach/11658#-KY-4NMjA-tl4jmSCct6, there's something to be said for all clusters going through the same process to get to where they are rather than old ones going through migrations and new ones being initialized directly in the desired state.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Comment by bdarnell** [comment]
Review status: 0 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/system.go, line 83 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYnu7YzOPUiIvrJU1qd:-KYqbOzTAp3EYFyliIR6:b-u1td91) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/sql/sqlbase/system.go#L83)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Agreed. I was going to actually remove this in a later PR (particularly since it was convenient for testing to still have the old schema around), but I suppose it makes the most sense to do it now.

However, as @bdarnell alluded to in https://reviewable.io/reviews/cockroachdb/cockroach/11658#-KY-4NMjA-tl4jmSCct6, there's something to be said for all clusters going through the same process to get to where they are rather than old ones going through migrations and new ones being initialized directly in the desired state.
</blockquote></details>

Yeah, we definitely want to be able to remove legacy code at some point. My assumption has been that we would remove the legacy code and the migration at the same time (so at some point after this migration has been "live" for a sufficient period of time (policy TBD), we'd "rebase" everything with all current migrations baked in). 

My first choice is to start up new clusters by bootstrapping them with legacy data and applying the migrations at startup, but this preference isn't strong.

If we do move to creating the table with the new schema from the beginning, then we should also write the migration key at the same time so that the migration doesn't run at all. Idempotence is a good idea but I'm not sure if we want to strictly require it for all migrations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

File: pkg/migrations/migrations.go [hunk 1]
```diff
@@ -232,6 +233,32 @@ func migrationKey(migration migrationDescriptor) roachpb.Key {
...
 }
+
+func eventlogUniqueIDDefault(ctx context.Context, r runner) error {
+	const alterStmt = "ALTER TABLE system.eventlog ALTER COLUMN uniqueID SET DEFAULT uuid_v4();"
+
+	// System tables can only be modified by a privileged internal user.
+	session := sql.NewSession(ctx, sql.SessionArgs{User: security.NodeUser}, r.sqlExecutor, nil, nil)
+	defer session.Finish(r.sqlExecutor)
+	var err error
+	for retry := retry.Start(retry.Options{MaxRetries: 5}); retry.Next(); {
```
Comment (vivekmenezes) [inline pkg/migrations/migrations.go]: Any particular reason you want to attempt this 5 times?

**Comment by a-robinson** [comment]
Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 255 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KY_6klskkc2x1MdcJ1U-r5-255:-KYt36PpFYP8OYnzSfdT:b-tl93e2) ([raw file](https://github.com/cockroachdb/cockroach/blob/580ebdc0fdbfdb11e1764082cc2cc7c20f63d2a5/pkg/migrations/migrations.go#L255)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Any particular reason you want to attempt this 5 times?
</blockquote></details>

I don't want to use the `base.DefaultRetryOptions` because I don't want it to retry forever, but it seems worth trying more than once. Should I be doing something else here?

---

*[pkg/sql/sqlbase/system.go, line 83 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYnu7YzOPUiIvrJU1qd:-KYt3JDG82KGaBVMEckv:ba9kpvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/sql/sqlbase/system.go#L83)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, we definitely want to be able to remove legacy code at some point. My assumption has been that we would remove the legacy code and the migration at the same time (so at some point after this migration has been "live" for a sufficient period of time (policy TBD), we'd "rebase" everything with all current migrations baked in). 

My first choice is to start up new clusters by bootstrapping them with legacy data and applying the migrations at startup, but this preference isn't strong.

If we do move to creating the table with the new schema from the beginning, then we should also write the migration key at the same time so that the migration doesn't run at all. Idempotence is a good idea but I'm not sure if we want to strictly require it for all migrations.
</blockquote></details>

In this particular case, the migration being run from uuid_v4 to uuid_v4 is practically the same as the migration from experimental_unique_bytes to uuid_v4. The `alter table` command still runs and updates the table descriptor. So this may be a case where we can reasonably remove the old bits at the same time as when we add the migration, because mechanically the same thing is being run.

I'll defer to you on this though. Want me to switch back to the initial version of this PR that kept using the experimental_unique_bytes function?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 255 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KY_6klskkc2x1MdcJ1U-r5-255:-KYu80BIaqxWZmcwFt7z:b3oecne) ([raw file](https://github.com/cockroachdb/cockroach/blob/580ebdc0fdbfdb11e1764082cc2cc7c20f63d2a5/pkg/migrations/migrations.go#L255)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I don't want to use the `base.DefaultRetryOptions` because I don't want it to retry forever, but it seems worth trying more than once. Should I be doing something else here?
</blockquote></details>

What do you fear about it retrying forever? If it retries and fails more than 5 times this will not update the DEFAULT value and we'll have a non-functional system when we remove experimental_unique_bytes() right?

Do we want to first check that the default value is not set correctly before we make an attempt to ALTER it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Comment by a-robinson** [comment]
Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 255 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KY_6klskkc2x1MdcJ1U-r5-255:-KYu9aMQxwVC-oj10uRa:b-8h4cz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/580ebdc0fdbfdb11e1764082cc2cc7c20f63d2a5/pkg/migrations/migrations.go#L255)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

What do you fear about it retrying forever? If it retries and fails more than 5 times this will not update the DEFAULT value and we'll have a non-functional system when we remove experimental_unique_bytes() right?

Do we want to first check that the default value is not set correctly before we make an attempt to ALTER it?
</blockquote></details>

If it retries forever, this process will hold the migration lease, preventing any other node from starting up even if it's just this node that's broken. If it returns an error, the node will kill itself.

ALTER TABLE is idempotent, so it seems fairly harmless to do it blindly vs adding extra code to try to detect whether the table is already in the desired state.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Comment by bdarnell** [comment]
Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/system.go, line 83 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYnu7YzOPUiIvrJU1qd:-KYvLwbm3w7uUh8zHt7p:buvqv0m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/sql/sqlbase/system.go#L83)):*
> So this may be a case where we can reasonably remove the old bits at the same time as when we add the migration, because mechanically the same thing is being run.

Good point. Yeah, let's just change the schema from the beginning here then.

My main concerns about running through the migrations versus bootstrapping everything in the new schema are A) tables created by bootstrapping are "special" in various ways and I'd rather not have new tables be special in this way on some clusters but not others, and B) not all ALTER TABLEs are idempotent. In this case it's easy to re-run the ALTER TABLE SET DEFAULT no matter what, but consider a future in which we have a series of migrations that create a table and later add columns and indexes to it. If we update the baseline each time then there are more possible permutations of previous state; it's simpler to run through all the migrations so each migration can assume that the previous migration has run and the next migration has not run (so the only thing it needs to check for idempotence is whether its own migration has been performed or not). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Comment by a-robinson** [comment]
Any other comments on this, or can we wrap it up tomorrow?

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/system.go, line 83 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KYnu7YzOPUiIvrJU1qd:-KYvRX-Pgxfg9ZWK3wF3:b4ggras) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ba91e09b273b374b407e807029096f58c309205/pkg/sql/sqlbase/system.go#L83)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> So this may be a case where we can reasonably remove the old bits at the same time as when we add the migration, because mechanically the same thing is being run.

Good point. Yeah, let's just change the schema from the beginning here then.

My main concerns about running through the migrations versus bootstrapping everything in the new schema are A) tables created by bootstrapping are "special" in various ways and I'd rather not have new tables be special in this way on some clusters but not others, and B) not all ALTER TABLEs are idempotent. In this case it's easy to re-run the ALTER TABLE SET DEFAULT no matter what, but consider a future in which we have a series of migrations that create a table and later add columns and indexes to it. If we update the baseline each time then there are more possible permutations of previous state; it's simpler to run through all the migrations so each migration can assume that the previous migration has run and the next migration has not run (so the only thing it needs to check for idempotence is whether its own migration has been performed or not). 
</blockquote></details>

Sounds like a plan.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-:-KYvRbPDvHAt8Q0Dyuoj:b-y4ijav)*

**Comment by bdarnell** [comment]
LGTM

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-:-KYvTB5RtYwaLlqRHGCm:bb6ywmx)*

**Comment by vivekmenezes** [comment]
:lgtm:

---

Review status: 0 of 18 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 255 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KY_6klskkc2x1MdcJ1U-r5-255:-KYxpE-OCnyTOqH3M-Sp:b2pdr82) ([raw file](https://github.com/cockroachdb/cockroach/blob/580ebdc0fdbfdb11e1764082cc2cc7c20f63d2a5/pkg/migrations/migrations.go#L255)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

If it retries forever, this process will hold the migration lease, preventing any other node from starting up even if it's just this node that's broken. If it returns an error, the node will kill itself.

ALTER TABLE is idempotent, so it seems fairly harmless to do it blindly vs adding extra code to try to detect whether the table is already in the desired state.
</blockquote></details>

Sounds fine, I missed the part about returning an error results in the node killing itself. Add a comment to that effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-:-KYxpdHiI-9gLttYaNCh:bnfp4nl)*

**Comment by a-robinson** [comment]
Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/migrations/migrations.go, line 255 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228#-KY_6klskkc2x1MdcJ1U-r5-255:-KYxyw4l8FDFVDiCdyp_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/580ebdc0fdbfdb11e1764082cc2cc7c20f63d2a5/pkg/migrations/migrations.go#L255)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Sounds fine, I missed the part about returning an error results in the node killing itself. Add a comment to that effect.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12228)*

**Pull request merged**